### DISCUSSION TITLE
Dynamic table names when using parameters in quarto

### DISCUSSION THREAD

**Discussion by nithinmkp** [quarto-dev/quarto-cli#11249] [Q&A]
### Description

Hello all,
I want to have dynamic table names (tables are based on computations, mostly in R) based on parameters using quarto document. How can this be done?

````{qmd}
---
title: analysis of `r params$cyl` cylinder cars
format: html
params:
  cyl : 4
---
We present analysis of `r params$cyl` cylinder cars

```{r}
library(tidyverse)
```
```{r}
# label: tbl-summary
# tbl-cap: sumamry of `r params$cyl` cylinder cars

mtcars |>
filter(cyl== params$cyl) |>
select(-cyl) |>
group_by(gear) |>
summarise(avg_mgp =mean (mpg,na.rm=TRUE))
```
````

**Comment by mcanouil** [comment]
What do you mean by "table names"?


The recommend way to use expression in captions is to use cross-ref div:

````

::: {#fig-amazing}

```{r}
whatever
```

A caption using `{r} params[["value"]]`
::: 

````


The legacy knitr approach:

````
```{r}
#| label: fig-amazing
#| fig-cap: !expr 'paste("A caption using", params[["value"]])'
whatever
```
````

In the future, please spend a bit more time searching through previous discussions:

- https://github.com/quarto-dev/quarto-cli/discussions/7308

**Comment by nithinmkp** [comment]
Thanks. But what about making the label dynamic too. I actually want to use
this for knitr::child_document or knutr::knit_expand

**Reply by mcanouil** [reply]
Use those functions to generate the whole code cell including the label.

**Reply by nithinmkp** [reply]
I could generate the captions. but dynamically referring them is not happening. For all tables, it is taking the last generated table number
This is the template i am using for child documents 
````{=qmd}
## Gapminder for {{country}}

::: {#tbl-gapminder_{{"country"}}}

```{r}
#| echo: false
gapminder |>
filter(continent == params$cont, 
       country == "{{country}}") |>
       pivot_longer(cols = -c(1:3)) |>
group_by(name) |>
summarize(mean = mean(value, na.rm = TRUE),
          sd = sd(value, na.rm = TRUE)) |>
gt()
```
Summary statistics of Gapminder dataset of {{country}}

:::

The `r glue::glue("@tbl-gapminder_{{"country"}}")` shows the summary statistics for  {{country}}  
````
main file
```{=qmd}
---
title: Gapminder data continent wise analysis- `r params$cont`
author:
  - Nithin M
date: '`r format(Sys.Date(), "%B %d, %Y")`'
format: 
  html: default
params:
  cont: Asia
  
fontsize: 12pt
---

# Country Wise Analysis
```{r}
#| results: hide
#| echo: false
library(gapminder)
library(tidyverse)
library(gt)
countries <- gapminder |>
             filter(continent == params$cont) |>
             distinct(country) |>
             pull()

expanded_child <- countries |> 
  map(function(country) knitr::knit_expand("temp2.qmd", country = country))  |> 
  flatten()

parsed_child <- knitr::knit_child(text = unlist(expanded_child))
```
`r parsed_child`
````

**Reply by mcanouil** [reply]
Could you check the formatting of your code?

> The `r glue::glue("@tbl-gapminder_{{"country"}}")` shows the summary statistics for  {{country}}  

This looks very weird, double quotes inside double quotes.

> date: '`r format(Sys.Date(), "%B %d, %Y")`'

You should not do this and actually have no reason to.
https://quarto.org/docs/reference/dates.html

**Comment by nithinmkp** [comment]
yes. dates was my bad. actually copied little bit from rmarkdown document.  What i really want to achieve is to refer to table generated from the child document. There will be a table div labbeled "gapminder_India" for example. How would i refer to this dynamically?

**Reply by mcanouil** [reply]
Your code is malformed so I can't use it, here is one that works:

````qmd
---
format: html
---

```{r}
#| include: false
for (ipkg in c("palmerpenguins", "gt", "dplyr")) {
  if (!require(ipkg, character.only = TRUE)) {
    install.packages(ipkg)
    library(ipkg, character.only = TRUE)
  }
}
```

Below are tables of the first six rows of the Palmer penguins dataset by species:

```{r}
#| echo: false
#| output: asis
cat(
  sprintf(
    "- `%s` (@tbl-%s)",
    levels(penguins[["species"]]),
    levels(penguins[["species"]])
  ),
  sep = "\n"
)

for (ispecies in levels(penguins[["species"]])) {
  tab <- penguins |>
    filter(species %in% ispecies) |>
    select(-species) |>
    head() |>
    gt() |>
    tab_header(title = ispecies)
  cat(sep = "\n", knitr::knit_child(quiet = TRUE, text = c(
    "```{r}",
    "#| echo: false",
    sprintf("#| tbl-cap: %s", ispecies),
    sprintf("#| label: tbl-%s", ispecies),
    "tab",
    "```"
  )))
}
```
````

Source: https://mickael.canouil.fr/posts/2023-03-05-quarto-auto-table-crossref/

**Comment by nithinmkp** [comment]
I dont know why it does not work for you, Thanks. 
This is the issue. While i refer, all tables get the same number 

![ss1](https://github.com/user-attachments/assets/1071aea9-01f0-4000-96ad-0d3fcfd84898)
![ss2](https://github.com/user-attachments/assets/a47fe797-0c53-457f-8211-74bddbe7fe46)

**Reply by mcanouil** [reply]
> I dont know why it does not work for you, Thanks.

Look at how formatted is your post/code.

As I showed, it works in Quarto, now it's for you to debug your R/knitr code.
You can use `keep-md: true` to see what is the intermediate markdown you are producing.

**Comment by nithinmkp** [comment]
Thanks and sorry. Figured that out, but i still cant manage when the names to pass have special characters between them
![Uploading ss3.png…]()

**Reply by mcanouil** [reply]
You need to clean those names as references should only be alphanumerical mostly ASCII.

**Comment by nithinmkp** [comment]
Thanks. That helped me solve it!!
